[Rocprofiler-systems] : Refactor papi enumeration to fix a hang on Intel systems

sputhala-amd opened the pull request:
## Motivation

Resolves: SWDEV-563681

When profiling with PAPI_enum_cmp_event() or initializing PAPI components, the application appeared to “hang” on Intel systems but completed normally on AMD EPYC systems.
PAPI uses libpfm4 and the Linux perf_event subsystem to enumerate available hardware counters from /sys/devices/.

On Intel system exposes a very large number of uncore performance monitoring units (PMUs) under /sys/devices/,
The PAPI perf_event_uncore component attempts to enumerate every one of these uncore PMUs and their associated events through libpfm.
On AMD EPYC systems, /sys/devices/ only contains a few entries such as amd_iommu_* and no uncore PMU directories, so enumeration completes almost instantly.

Enumerating through this large number of perf_event_uncore events cause a hang like situation for a long time.

## Technical Details

- Add an exclude argument to available_events_info() for perf_event_uncore causing hang like case on Intel systems with large number of uncore events.
- Enumerate papi available events only when papi events are specified by users inside early initialization logic
- Move papi available event query for ROCPROFSYS_SAMPLING_OVERFLOW_EVENT config setting to the avail component, to move the heavy logic outside initialization.
- Make category option for rocprof-sys-avail -H -c case insensitive
- Sort the output of the hardware counter output alphabetically.
- Provide new option to query available overflow events that can be specified for ROCPROFSYS_SAMPLING_OVERFLOW_EVENT using new command option rocprof-sys-avail -H -c overflow

## Test Plan

Manual tests performed.

## Test Result

Expected test results achieved

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

sputhala-amd committed:
Refactor papi enumeration to fix a hang on Intel systems

- Add an exclude argument to available_events_info() for
  perf_event_uncore causing hang like case on Intel systems with large
number of uncore events.
- Enumerate papi available events only when papi events are specified by
  users inside early initialization logic
- Move papi available event query for ROCPROFSYS_SAMPLING_OVERFLOW_EVENT
  config setting to the avail component, to move the heavy logic outside
initialization.
- Make category option for rocprof-sys-avail -H -c case insensitive
- Provide new option to query available overflow events that can be
  specified for ROCPROFSYS_SAMPLING_OVERFLOW_EVENT using new command
option rocprof-sys-avail -H -c overflow

sputhala-amd commented:
This PR needs to be merged after the timemory PR is merged: https://github.com/ROCm/timemory/pull/13

mradosav-amd commented on projects/rocprofiler-systems/source/bin/rocprof-sys-avail/common.cpp:
```diff
@@ -307,30 +308,59 @@ process_categories(parser_t& p, const str_set_t& _category_options)
     std::vector<std::function<void()>> _shorthand_patches{};
+
+    // Helper to do case-insensitive string comparison
+    auto _tolower = [](std::string str) {
```
```suggestion
    auto _tolower = [](const std::string& str) {
```

sputhala-amd committed:
Update projects/rocprofiler-systems/source/bin/rocprof-sys-avail/common.cpp

Co-authored-by: Milan Radosavljevic <milan.radosavljevic@amd.com>

dgaliffiAMD committed:
Update timemory submodule pointer

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

sputhala-amd committed:
Fix errors on compile

sputhala-amd commented:
@mradosav-amd since you reviewed this PR before would you be able to review the PR?

marantic-amd commented on projects/rocprofiler-systems/source/bin/rocprof-sys-avail/common.cpp:
```diff
@@ -307,30 +308,60 @@ process_categories(parser_t& p, const str_set_t& _category_options)
+    };
+
+    // Helper to find case-insensitive match in category options
+    auto _find_category = [&_category_options,
```
You could optimize this with caching somewhere the lowercase categories. Since now `_find_category ` lambda is creates a new string copy for every comparison. Some quick math, for around 30 categories you would have made more than 500 string copies. Utilizing `unordered_map` could be some simple solution, like this:

```suggestion
    std::unordered_map<std::string, std::string> _category_map;
    for(const auto& opt : _category_options)
    {
        _category_map[_tolower(opt)] = opt;
    }
```

And then the `_find_category` could look like this:
```suggestion
    // Helper to find case-insensitive match in category options
    auto _find_category = [&_category_map,
                           &_tolower](const std::string& input) -> std::string {
        auto input_lower = _tolower(input);
        auto it = _category_map.find(input_lower);
        return (it != _category_map.end()) ? it->second : "";
    };
```

sputhala-amd replied:
Agree. Made the changes.

mradosav-amd commented on projects/rocprofiler-systems/source/lib/core/config.cpp:
```diff
@@ -942,12 +943,17 @@ configure_settings(bool _init)
         {
-            if(itr.available()) _papi_choices.emplace_back(itr.symbol());
+            std::vector<std::string> _papi_choices = {};
+            for(auto itr : tim::papi::available_events_info({ "perf_event_uncore" }))
```
```suggestion
            for(const auto& itr : tim::papi::available_events_info({ "perf_event_uncore" }))
```

mradosav-amd commented on projects/rocprofiler-systems/source/bin/rocprof-sys-avail/avail.cpp:
```diff
@@ -1097,10 +1110,11 @@ write_hw_counter_info(std::ostream& os, const array_t<bool, N>& options,
-    auto fields =
-        std::vector<std::pair<std::string, hwcounter_info_t>>{ { "CPU", _papi_events },
-                                                               { "GPU", _rocm_events } };
+    auto fields = std::vector<std::pair<std::string, hwcounter_info_t>>{
```
```suggestion
    const auto fields = std::vector<std::pair<std::string, hwcounter_info_t>>{
```

sputhala-amd replied:
This change to make fields const would break since it's being modified in the for loop later.

mradosav-amd commented on projects/rocprofiler-systems/source/bin/rocprof-sys-avail/common.cpp:
```diff
@@ -307,30 +308,60 @@ process_categories(parser_t& p, const str_set_t& _category_options)
+        for(const auto& opt : _category_options)
+        {
+            if(_tolower(opt) == input_lower) return opt;
+        }
```
Can you iterate over _category_options, convert it to lower and store it? Then you can just call `std::find`?
This is converting **all** categories to lower on **each** invocation.

mradosav-amd commented on projects/rocprofiler-systems/source/bin/rocprof-sys-avail/common.cpp:
```diff
@@ -307,30 +308,60 @@ process_categories(parser_t& p, const str_set_t& _category_options)
     for(const auto& itr : category_view)
     {
-        auto _is_shorthand = [&_shorthand_patches, &_category_options,
+        auto _is_shorthand = [&_shorthand_patches, &_find_category,
```
I guess there is no need to instantiate lambda each time in for loop.

mradosav-amd commented on projects/rocprofiler-systems/source/bin/rocprof-sys-avail/common.cpp:
```diff
@@ -307,30 +308,60 @@ process_categories(parser_t& p, const str_set_t& _category_options)
+
+    // Helper to find case-insensitive match in category options
+    auto _find_category = [&_category_options,
+                           &_tolower](const std::string& input) -> std::string {
```
Can we switch to `std::string_view`? It will prevent unnecessary copy or string creation.

mradosav-amd commented on projects/rocprofiler-systems/source/bin/rocprof-sys-avail/common.cpp:
```diff
@@ -307,30 +308,60 @@ process_categories(parser_t& p, const str_set_t& _category_options)
-                          "valid categories");
+            _shorthand_patches.emplace_back([itr, _matched]() {
+                category_view.erase(itr);
+                category_view.emplace(_matched);
```
Are `itr` and `_matched` same values or `_matched` is only `to_lower(itr)`?

sputhala-amd replied:
itr seems to be what the user passed and _matched is what the available category options are as populated from inside avail.cpp.

marantic-amd reviewed (changes requested):
Please try to optimize the implementation, since we are trying not to insert any additional overhead

mradosav-amd reviewed (changes requested):
Thank you @sputhala-amd. Here are some suggestions.

sputhala-amd committed:
Change 1: Optimization for the category matching lambda

Optmization changes.

sputhala-amd commented:
@marantic-amd @mradosav-amd Thanks both for the review. I have addressed all but one suggested change.

marantic-amd reviewed (approved):
LGTM

dgaliffiAMD commented on projects/rocprofiler-systems/source/bin/rocprof-sys-avail/avail.cpp:
```diff
@@ -1097,10 +1110,11 @@ write_hw_counter_info(std::ostream& os, const array_t<bool, N>& options,
-        std::vector<std::pair<std::string, hwcounter_info_t>>{ { "CPU", _papi_events },
-                                                               { "GPU", _rocm_events } };
+    auto fields = std::vector<std::pair<std::string, hwcounter_info_t>>{
+        { "CPU", _papi_events }, { "GPU", _rocm_events }, { "overflow", _overflow_events }
```
I think that you also need to increase the width of the "device" column. 

Something like this - 
https://github.com/ROCm/rocm-systems/blob/5d43b289fdb665922e5f23a8ad655329a9746e02/projects/rocprofiler-systems/source/bin/rocprof-sys-avail/avail.cpp#L1191-L1192

```
width_type _w = { { (int64_t) itr.symbol().length(), (int64_t) 8, (int64_t) 6,
```

sputhala-amd committed:
Modify the rocprof-sys-avail -c option for overflow

Overflow should not be displayed as a device in rocprof-sys-avail -H -c CPU

Users can instead do regex on summary where overflow is appended in description

User can do rocprof-sys-avail -H -c CPU -d -r overflow

dgaliffiAMD committed:
Revert change to column width

dgaliffiAMD commented:
LGTM. I just rebasing.
